### PR TITLE
Emitters and Shieldgens work roundstart, APCs charge evenly

### PR DESCRIPTION
**Pull request by MTandi**
## About The Pull Request

Fixed the issue when the power priority was preferring random machinery around the station over the crucial Emitters and Shieldgens.

Because of how the consumers are selected right now, PACMAN attached to a station network will give power for the machinery/lights/environment of a random APC instead of the emitters next to it.

This also makes power sinks process before APCs.

Also made APCs charge in cascades within a powernet:

1. APCs try to charge to 5% to enable environment channel.
2. When all are above 5%, then allow APCs to charge to 20% to enable lighting channel
2. When all are above 20%, then allow APCs to charge to 35% to enable equipment channel
3. When all are above 35%, then allow APC to charge to full battery

As a result, a low-rate power source can provide enough power to at least make all airlocks working and have dim lights enabled.

Charlie station APCs after you run PACMAN with 15 uranium sheets:
image

## Why It's Good For The Game

Makes emitters actually work, instead of having people fire kisses at SM to raise the internal energy.
Makes it possible to use PACMAN to kickstart SM.
Gives a reason to set up solars.
Makes the power distribution more fair and equal after the recent APC battery buff.

## Changelog

:cl:
fix: Fixed emitters, shield gens and other wired machinery having lower power consumption priority than APCs
qol: Made APCs charge more evenly to ensure that environment and lights are working everywhere before giving the power for equipment
/:cl:

### CODE REVIEW

**Comment by MTandi** [comment]
> APC charging isn't even handled in apc early processing. It's handled in late processing which happens after every other machine processes. This PR fixes absolutely nothing. When you were talking about emitters not firing, you showed orange emitters while SMES were set. Orange emitters means there is 0 power at all, not because there's not enough excess. This implies a rat bit the cables in maintenance, disconnecting the SMES from the grid.

Run meta station from my branch vs. master. 
On master you click on emitters, they don't fire. 
Here you click on emitters, they fire.

File: code/controllers/subsystem/machines.dm [hunk 1]
```diff
@@ -82,6 +82,21 @@ SUBSYSTEM_DEF(machines)
 	if (!resumed)
 		for(var/datum/powernet/powernet as anything in powernets)
 			powernet.reset() //reset the power state.
+		current_part = SSMACHINES_MACHINES
+		src.currentrun = processing.Copy()
+
+	//Processing machines, directly wired to the network.
```
Comment (Pickle-Coding) [inline code/controllers/subsystem/machines.dm:88]: This is inaccurate. This covers every single machine that will get processed and aren't in the other lists.

**Comment by lessthnthree** [comment]
Wouldn't leaving the emitters in the normal grid connected machines pool still leave the possibility other machinery gets powernet power before the emitters?

Emitters should be in a group that get to pull power before other machinery.

**Comment by MTandi** [comment]
> Wouldn't leaving the emitters in the normal grid connected machines pool still leave the possibility other grid-connected machinery gets powernet power before the emitters?
> 
> Emitters should be in a group that get to pull power before other machinery.

It would, but roundstart there is nothing else on the station grid except for emitters and shieldgens. This is really just another bandaid fix, we need to redo the power consumption logic.

Charlie station is broken right now.

**Comment by lessthnthree** [comment]
> > Wouldn't leaving the emitters in the normal grid connected machines pool still leave the possibility other grid-connected machinery gets powernet power before the emitters?
> > Emitters should be in a group that get to pull power before other machinery.
> 
> It would, but roundstart there is nothing else on the station grid except for emitters and shieldgens. This is really just another bandaid fix, we need to redo the power consumption logic.
> 
> Charlie station is broken right now.

Yeah, I'm openly musing if this could while we're here be adapted to also help emitters losing power midround because the grid doesn't have enough power.

**Comment by MTandi** [comment]
Drafting. I will redo this to cover all powernet related issues, not just emitters not working.

**Comment by MTandi** [comment]
Meta station on solars on this branch:

image

Enough to power all environment and lights plus about 40% of equipment on the roundstart setup

**Comment by Pickle-Coding** [comment]
Don't use the bridge power monitoring consoles. They're bugged. Use the ampcheck on the console in engineering for a more accurate reading. I'll properly review this later.

**Comment by Pickle-Coding** [comment]
Why is APC early processing getting processed after the channel charging? This is causing some APCs to get discharged during the channel charging phases. I also noticed APCs aren't updating their icons to turn green. Once an APC turns red (which will be all of them at round start), then they're going to be red for the entire round.

**Comment by MTandi** [comment]
> Why is APC early processing getting processed after the channel charging? This is causing some APCs to get discharged during the channel charging phases. I also noticed APCs aren't updating their icons to turn green. Once an APC turns red (which will be all of them at round start), then they're going to be red for the entire round.

Because charging works on threshold basis, the charging happens before the static usage to enable certain channels.

For instance, when the cell has 14%, we charge it to 15%, and this will enable the lighting channel.

The process may drain this back to 14%, but then we charge it back to 15% again on the next fire.

This way, when the station is low on power, you will have all station APCs being charged slightly above the tresholds (5% buffer to avoid flickering), as on my example with solars.

As for the icon, the charging state is properly staying at `APC_FULLY_CHARGED`, so it's just the icon issue. You can see the icons turning green in UI I'll make a hotfix for the icon update.

NVIDIA_Overlay_oN10N4b48T

**Comment by Pickle-Coding** [comment]
Alright I understand what you're doing a bit more now. I think it would be less confusing if APCs that charge to remain at the same charge to upkeep a channel shouldn't show as charging though (both console and its icon). Also that specific charging shouldn't count towards the area's power usage reporting, since that double counts the energy used to cause the discharge. 

This will communicate better to the players that it's just the higher charged APCs discharging first and not just weird charging discharging phenomena.

**Pull request merged**